nixos module system: enhance type-checking

ts468 opened the pull request:
Add the tool "nixos-typecheck" that can check an option declaration to:
- Enforce that an option declaration has a "defaultText" if and only if the
  type of the option derives from "package", "packageSet" or "nixpkgsConfig"
  and if a "default" attribute is defined.
- Enforce that the value of the "example" attribute is wrapped with "literalExample"
  if the type of the option derives from "package", "packageSet" or "nixpkgsConfig".
- Warn if a "defaultText" is defined in an option declaration if the type of
  the option does not derive from "package", "packageSet" or "nixpkgsConfig".
- Warn if no "type" is defined in an option declaration.

Remove all errors that were found by "nixos-typecheck".

ts468 commented:
@nbp Master of the module system ;), do you have any objections against this PR?

@edolstra Could you maybe check the rules for option declarations that I wrote into the manual?

Already, thanks for your help! :)

domenkozar commented:
@ts468 all changes that add a type could be committed separately without a review, since that's not controversial.

ts468 commented:
Sorry, it seems like I was too quick, there is still an error in it that I have to fix.

ts468 commented:
Now it should work! :)

nbp commented on lib/types.nix:
```diff
@@ -181,81 +247,165 @@ rec {
+    # Complex types
+    #
+
+    functionTo = resultType:
```
Remove this one, no function should be allowed in the module system!
This should be forbidden by design, unfortunately `nixpkgs.config` uses it :(

nbp commented on lib/types.nix:
```diff
@@ -181,81 +247,165 @@ rec {
+    #
+
+    functionTo = resultType:
+      mkOptionType {
```
Remove this `functionTo` type.

Function should be forbidden by default (even if `nixpkgs.config` use one) because they are impossible to document properly, and to introspect statically.

nbp commented on lib/types.nix:
```diff
@@ -181,81 +247,165 @@ rec {
       };
 
-    enum = values:
+    nixpkgsConfig = pkgs:
```
I do not see any reason to move this type into this file.  As the name implies, this is the type of this unique option.

nbp commented on lib/modules.nix:
```diff
@@ -424,12 +435,66 @@ rec {
     in
       if opt.type.getSubModules or null == null
-      then opt // { type = f (opt.type or types.unspecified); }
+      then opt // { type = f (opt.type or (inferType typeInference loc opt)); }
```
I would prefer the type enforcement to be mandatory, we a clear release note message to explain the steps to NixOS developer on how they can change their modules to use `types.unspecified`.

As this is a non-backward compatible change, this should probably be a warning first in 16.03, and enforced in the next version of NixOS.  Having extra types is backward compatible, so I do not see the need for `_module.typeInference`.

ts468 replied:
@nbp My thought was that `_module.typeInference` could be used to disable the type inference, and so a performance hit, for an ordinary evaluation. Do you think that's not necessary?

nbp commented on lib/types.nix:
```diff
@@ -181,81 +247,165 @@ rec {
+            internalModule = [ { inherit _module; } ];
             coerce = def: if isFunction def then def else { config = def; };
-            modules = opts' ++ map (def: { _file = def.file; imports = [(coerce def.value)]; }) defs;
+            modules = opts' ++ internalModule ++ map (def: { _file = def.file; imports = [(coerce def.value)]; }) defs;
```
We do not want to inherit from modules, especially if we have named submodules of named submodules. In such case, we would have a conflicting definitions of `_module.args.name`.

ts468 replied:
@nbp What would a better way to pass `_module.args`, `_module.check` and `_module.typeInference` into the evaluation of a submodule? Any suggestions?

nbp commented:
Split this commit in multiple branches, one for fixing the current tree, one for adding a warning for the lack of types.

Also note that the current commit add a lot of code into the evaluation of every option while not giving any value at the evaluation, but only at the execution of the `nixos-typecheck` command.  What you can do instead, is to collect all option declarations and check if they have a defined `type` attribute.  

Have a look at how we generate the documentation, we collect the list of option declarations to inject, the name of the type in the documentation.

ts468 commented:
@nbp Nicolas, thank you very much for your review, and sorry for my late reply! I'll work through your comments, they are very helpful! Thanks!

ts468 commented:
@domenkozar You're right, I committed them separately just right now ;)

Commit:
Add the tool "nixos-typecheck" that can check an option declaration to:

 - Enforce that an option declaration has a "defaultText" if and only if the
   type of the option derives from "package", "packageSet" or "nixpkgsConfig"
   and if a "default" attribute is defined.

 - Enforce that the value of the "example" attribute is wrapped with "literalExample"
   if the type of the option derives from "package", "packageSet" or "nixpkgsConfig".

 - Warn if a "defaultText" is defined in an option declaration if the type of
   the option does not derive from "package", "packageSet" or "nixpkgsConfig".

 - Warn if no "type" is defined in an option declaration.

ts468 commented:
Pushed as cad8957eabcbf73062226d28366fd446c15c8737
